### PR TITLE
[client-cpp] add subscription properties to consumer for cpp

### PR DESCRIPTION
**Pull request by Lannnnh**
# Motivation
Pulsar already support entry filter, But pulsar-client-cpp ConsumerImpl not contains subscriptionProperties Option. So, this PR want to solve it.

# Modifications
Enable subscriptionProperties on consumer for cpp client.

# Results
subscriptionProperties are added for CommandSubscribe. However, there is no way to verify the consumer subscriptionProperties. so I didn't add any specific tests.

# Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 
  
- [x] `no-need-doc` 
  
Why?
this PR only involves pulsar-client-cpp consumer subscribe opition modification

### CODE REVIEW

File: pulsar-client-cpp/lib/Commands.cc [hunk 1]
```diff
@@ -308,6 +309,14 @@ SharedBuffer Commands::newSubscribe(const std::string& topic, const std::string&
         subscribe->mutable_metadata()->AddAllocated(keyValue);
     }
 
+    for (std::map<std::string, std::string>::const_iterator it = subscriptionProperties.begin();
+         it != subscriptionProperties.end(); it++) {
```
Comment (BewareMyPower) [inline pulsar-client-cpp/lib/Commands.cc:313]: I think we can use the C++11 style for new code. Though some legacy code still uses older style. We can do some refactoring to these legacy code later.

Comment (Lannnnh) [reply pulsar-client-cpp/lib/Commands.cc]: thanks for your suggestion, it has been revised.

File: pulsar-client-cpp/lib/ConsumerConfiguration.cc [hunk 2]
```diff
@@ -214,6 +214,19 @@ ConsumerConfiguration& ConsumerConfiguration::setProperties(
...
+    return impl_->subscriptionProperties;
+}
+
+ConsumerConfiguration& ConsumerConfiguration::setSubscriptionProperties(
+    const std::map<std::string, std::string>& subscriptionProperties) {
+    for (std::map<std::string, std::string>::const_iterator it = subscriptionProperties.begin();
+         it != subscriptionProperties.end(); it++) {
+        impl_->subscriptionProperties.insert(std::make_pair(it->first, it->second));
+    }
+    return *this;
```
[the remark this answers is not in the document]
Comment (Lannnnh) [reply pulsar-client-cpp/lib/ConsumerConfiguration.cc]: It has been revised  as the previous question.

**Comment by Lannnnh** [comment]
>

**Comment by Lannnnh** [comment]
@BewareMyPower ok, I have been added a test in ConsumerConfigurationTest.

File: pulsar-client-cpp/lib/ConsumerConfiguration.cc [hunk 3]
```diff
@@ -214,6 +214,19 @@ ConsumerConfiguration& ConsumerConfiguration::setProperties(
...
 
+std::map<std::string, std::string>& ConsumerConfiguration::getSubscriptionProperties() const {
+    return impl_->subscriptionProperties;
+}
+
+ConsumerConfiguration& ConsumerConfiguration::setSubscriptionProperties(
+    const std::map<std::string, std::string>& subscriptionProperties) {
+    for (const auto& subscriptionProperty : subscriptionProperties) {
+        impl_->subscriptionProperties.emplace(
+            std::make_pair(subscriptionProperty.first, subscriptionProperty.second));
```
Comment (BewareMyPower) [inline pulsar-client-cpp/lib/ConsumerConfiguration.cc:225]: There is no need to emplace a `std::pair`. See following example:

```c++
std::map<std::string, std::string> m;
m.emplace("key", "value");
```

Comment (Lannnnh) [reply pulsar-client-cpp/lib/ConsumerConfiguration.cc]: ok, it has been revised, thanks for your attention.

**Review by BewareMyPower — APPROVED** [review]

**Comment by Lannnnh** [comment]
@BewareMyPower Hello, this PR has been suspended for a few days. Next, how to proceed further?

**Comment by BewareMyPower** [comment]
It's blocked by the failed CI caused by some flaky tests. I reran the failed CI again just now.

image

BTW, you can also comment `/pulsarbot run-failure-checks` to rerun the tests. For example, https://github.com/apache/pulsar/pull/15081#issuecomment-1092507443

**Comment by BewareMyPower** [comment]
If it still failed, you can try to rebase to master and push by force again because the latest commits might fix some flaky tests.

**Comment by Lannnnh** [comment]
@BewareMyPower Thanks for the reminder, now that all tests have passed. Next, need your review/accept to merge this PR.

**Pull request merged**